### PR TITLE
Stack id validation

### PR DESCRIPTION
**Pull request by matthewmcnew**
- Don't create CustomBuilder's if stack is not ready. 
- No mixin validation, yet.

### CODE REVIEW

File: pkg/cnb/builder_builder.go [hunk 1]
```diff
@@ -172,7 +171,21 @@ func (bb *BuilderBuilder) writeableImage() (v1.Image, error) {
 	})
 }
 
-func (bb *BuilderBuilder) lifecycleLayer() (v1.Layer, error) {
+func (bb *builderBlder) validateBuilder() {
+	for _, bp := range bb.buildpackLayers {
+		err := bp.BuildpackLayerInfo.supports(bb.stackId)
+		if err != nil {
+			bb.err = errors.Wrapf(err, "validating buildpack: %s", bp.BuildpackInfo)
+			return
```
Comment (ashwin-venkatesh) [inline pkg/cnb/builder_builder.go]: i wonder if we should go through every buildpack in the spec and list all the once that are unsupported

File: pkg/cnb/create_builder.go [hunk 2]
```diff
@@ -34,11 +34,17 @@ func (r *RemoteBuilderCreator) CreateBuilder(keychain authn.Keychain, buildpackR
...
-	builderBuilder, err := newBuilderBuilder(buildImage, lifecycleImage, stack, r.KpackVersion)
+	builderBldr, err := newBuilderBldr(lifecycleImage, r.KpackVersion)
 	if err != nil {
 		return v1alpha1.BuilderRecord{}, err
 	}
 
+	builderBldr.AddStack(
+		stack.Spec.Id,
+		buildImage,
+		[]string{},
```
Comment (ashwin-venkatesh) [inline pkg/cnb/create_builder.go]: one day, we shall have mixins :)

File: pkg/reconciler/v1alpha1/custombuilder/custombuilder.go [hunk 3]
```diff
@@ -121,6 +123,10 @@ func (c *Reconciler) reconcileCustomBuilder(customBuilder *expv1alpha1.CustomBui
 		return v1alpha1.BuilderRecord{}, err
 	}
 
+	if !stack.Status.GetCondition(corev1alpha1.ConditionReady).IsTrue() {
```
Comment (ashwin-venkatesh) [inline pkg/reconciler/v1alpha1/custombuilder/custombuilder.go]: ❤️

File: pkg/cnb/builder_builder.go (hunk 1 above)
Comment (matthewmcnew) [reply pkg/cnb/builder_builder.go]: That would require us to combine the errors. Which would be great :)

File: pkg/cnb/create_builder.go (hunk 2 above)
Comment (matthewmcnew) [reply pkg/cnb/create_builder.go]: Maybe, we could remove this field until we support it. I was just playing with what it looks like.

File: pkg/cnb/builder_builder.go [hunk 4]
```diff
@@ -35,48 +35,46 @@ const (
 
 var normalizedTime = time.Date(1980, time.January, 1, 0, 0, 1, 0, time.UTC)
 
-type BuilderBuilder struct {
+type builderBlder struct {
+	err error
```
Comment (matthewmcnew) [inline pkg/cnb/builder_builder.go]: Error handling strategy inspired by Dave Cheney 

https://dave.cheney.net/2019/01/27/eliminate-error-handling-by-eliminating-errors

File: pkg/cnb/create_builder.go (hunk 2 above)
Comment (ashwin-venkatesh) [reply pkg/cnb/create_builder.go]: It threw me off a little bit when I saw it getting passed in and had to do a double take. I'm alright with keeping it around if we throw a tiny comment about why we are passing in an empty slice 🙂

Comment (matthewmcnew) [reply pkg/cnb/create_builder.go]: Mixins are the next thing to work on :)

**Pull request merged**